### PR TITLE
Implementing inference

### PR DESCRIPTION
**Pull request by rbevansp**
Use Typer cli arguments to specify from which run you wish to take the model to perform inference
Refactor the inference code
Get the inference code to first print a summary of the experiment name and hyperparameters that were used to create that run
Print the model’s predicted label and prediction confidence in a nice way

### CODE REVIEW

File: bin/cli.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 from datetime import datetime
 from pathlib import Path
+from typing_extensions import Annotated
```
Comment (e-vissch) [inline bin/cli.py:3]: As you ended up not using this, you should delete this. VS Code has some automated tools to remove unused imports, and it should show up as grey if you don't use it. 

Note the reason that it didn't work originally is because the documentation for Argument is from a later version than what we were using.

File: bin/cli.py [hunk 2]
```diff
@@ -43,6 +44,7 @@ def train(
     fmt = "%Y-%m-%dT%H-%M"
     timestamp = datetime.strftime(datetime.utcnow(), fmt)
     run_path = RESULTS_DIR / name / timestamp
+    print(run_path)
```
Comment (e-vissch) [inline bin/cli.py:47]: is this leftover from debugging or did you intend it to be here.

File: ser/inference.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    confidence = max(list(torch.exp(output)[0]))
+    pixels = images[0][0]
+
+    # Generate ASCII art representation
+    ascii_art = generate_ascii_art(pixels)
+
+    print(f"Predicted Label: {pred} | Prediction Confidence: {confidence:.2%}")
+
+
+def generate_ascii_art(pixels):
```
Comment (e-vissch) [inline ser/inference.py:48]: you could create a separate utils file for this stuff, given it's not explicitly related to inference.

File: ser/inference.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        params_data = json.load(f)
+    params = Params(**params_data)
+
+    # Print experiment summary
+    print("Experiment Name:", params.name)
+    print("Hyperparameters:")
+    print("\tEpochs:", params.epochs)
+    print("\tBatch Size:", params.batch_size)
+    print("\tLearning Rate:", params.learning_rate)
+
```
Comment (e-vissch) [inline ser/inference.py:25]: You are mixing what we call high and low-level logic here. Ideally in your cli file- the entry point- we can see the high level logic- load the params, print the params, get the data, get an example of the image with label, get the model then run the inference- the low level details. More abstraction would be good here, and some of this code should stay in the main cli file.

File: ser/inference.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import torch
+from pathlib import Path
+from ser.data import test_dataloader
+from ser.transforms import transforms, normalize
+from ser.params import Params
+import json
+
+def run_infer(run_path: str):
+    label = 6
+
```
Comment (e-vissch) [inline ser/inference.py:12]: Ideally the label is not hardcoded but something passed in by the user. Maybe they want an example of an image of number 3 or 8 etc.

File: ser/inference.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    # Select image to run inference for
+    dataloader = test_dataloader(1, transforms(normalize))
+    images, labels = next(iter(dataloader))
+    while labels[0].item() != label:
+        images, labels = next(iter(dataloader))
+
+    # Load the model
+    model = torch.load(Path(run_path) / "model.pt")
+
+    # Run inference
```
Comment (e-vissch) [inline ser/inference.py:35]: here is the inference- low level code

**Review by e-vissch — COMMENTED** [review]
Nice job! Make sure to look at the solutions as this might make some comments clearer. Lmk if you have any qs!
